### DISCUSSION TITLE
How to emit haptic feedback ?

### DISCUSSION THREAD

**Discussion by Louis-PAGNIER** [skiptools/skip#133] [Q&A] [answered]
Hello,

I tried many things to emit haptic feedback on android side but without any success :(
In the following example, the function must be called from a compose context but I want to call it from any context, like a simple button.
Does someone have an idea on how to fix it ?

```Swift
func feedback() {
    #if !SKIP
    UIImpactFeedbackGenerator(style: .medium).impactOccurred()
    #else
    // Need to be inside a @Composable
    let view = LocalView.current
    view.performHapticFeedback(HapticFeedbackConstantsCompat.KEYBOARD_TAP)
    #endif
}

```

**Comment by marcprux** [comment]
Untested, but how about something like this:

```swift
#if !SKIP
UIImpactFeedbackGenerator(style: .medium).impactOccurred()
#else
let context = ProcessInfo.processInfo.androidContext // Android-specific extension to get the global Context
// https://developer.android.com/reference/android/os/VibratorManager
if let vibratorManager = context.getSystemService(android.content.Context.VIBRATOR_SERVICE) as? android.os.VibratorManager {
    // https://developer.android.com/reference/android/os/Vibrator
    if let vibrator = vibratorManager.getDefaultVibrator() {
        // https://developer.android.com/reference/android/os/VibrationEffect
        vibrator.vibrate(android.os.VibrationEffect.createPredefined(android.os.VibrationEffect.EFFECT_CLICK))
    }

}
#endif
```

**Reply by Louis-PAGNIER** [reply]
Thank you !

It compile, but as I don't currently have a physical Android device, I can't confirm with the emulator alone.
I will test it in the future, but if someone has time to test the snippet, it is welcome.

**Comment by marcprux** [comment, marked as answer]
This seemed like it would be useful feature, so I've gone ahead and added the API (`UIImpactFeedbackGenerator`, `UINotificationFeedbackGenerator`, and `UISelectionFeedbackGenerator`) to skip-ui 0.9.5. Once you update with `File`/`Packages`/`Update to Latest Package Versions`, you should be able to just use those types to create haptic feedback on the Android side in the same way as the iOS side.

Note that you will need to add the `android.permission.VIBRATE ` permission to your `AndroidManifest.xml` file. See: https://skip.tools/docs/modules/skip-ui/#haptics

Please let us know if you have any questions or problems with the new feature, and thanks for the suggestion!
